### PR TITLE
💄 Refactor the coveragepy config

### PR DESCRIPTION
**Pull request by webknjaz**
This is a combination of changes that include:
- setting up local coveragepy HTML output to show context
- setting up mapping between site-packages and the repo
- measuring all the Python files in the repo
- including Cython
- making xfails not influence the coverage metric (these test functions may start executing different amounts of lines over time, causing random coverage drops/gains)
- using relative paths for playing nice with Codecov

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -268,6 +268,20 @@ directory = "misc"
...
 
+[tool.coverage.html]
+show_contexts = true
+skip_covered = false
+
+[tool.coverage.paths]
+_site-packages-to-src-mapping = [
+  "src",
+  "*/src",
+  '*\src',
```
Comment (CoolCat467) [inline pyproject.toml:279]: Why does it have both forward and backwards slash?

**Review by CoolCat467 — APPROVED** [review]
Looks good as far as I can tell, just a question about one of the config changes

File: pyproject.toml (hunk 1 above)
Comment (CoolCat467) [reply pyproject.toml:279]: also one below this also has a backwards slash

Comment (webknjaz) [reply pyproject.toml:279]: Probably not, but I saw this practice a lot in the wild. I think modern coveragepy normalizes them but I don't want to take a chance.

**Comment by webknjaz** [comment]
Yeah, I only had to adapt my typical dedicated config (which is easier to maintain) to the toml combo. It's all battle-tested, otherwise.

File: pyproject.toml (hunk 1 above)
Comment (CoolCat467) [reply pyproject.toml:279]: Probably not what?

Comment (webknjaz) [reply pyproject.toml:279]: Oh, my brain parsed it as “is this necessary?” :laughing: Apologies.

These are basically explicit Windows path variants.

Comment (CoolCat467) [reply pyproject.toml:279]: Ah, that makes a lot of sense.

File: pyproject.toml [hunk 2]
```diff
@@ -282,10 +296,15 @@ omit = [
 # The test suite spawns subprocesses to test some stuff, so make sure
 # this doesn't corrupt the coverage files
 parallel = true
+plugins = []
+relative_files = true
+source = ["."]
```
Comment (graingert) [inline pyproject.toml:301]: Why do we need source and source_pkgs? We should only need one

File: tests/cython/run_test_cython.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from . import test_cython  # noqa: F401
```
Comment (graingert) [inline tests/cython/run_test_cython.py:1]: Possibly it makes sense to refactor test_cython to export a main function and we can then call it here

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -234,15 +234,34 @@ jobs:
...
-        run: python -c 'import tests.cython.test_cython'
+        run: coverage run -m tests.cython.run_test_cython
+
+      - name: get Python version for codecov flag
+        id: get-version
+        run: >-
+          echo "version=$(python -V | cut -d' ' -f2 | cut -d'.' -f1,2)"
+          >> "${GITHUB_OUTPUT}"
+      - if: always()
+        uses: codecov/codecov-action@v5
```
Comment (jakkdl) [inline .github/workflows/ci.yml:260]: is there any problem with mixing versions? I previously hit errors I thought was caused by that.

File: pyproject.toml (hunk 1 above)
Comment (jakkdl) [reply pyproject.toml:279]: Having recently skimmed a bunch of codecov docs I'm pretty sure there's enough path normalization mentioned in a bunch of places that this shouldn't be needed.

File: pyproject.toml (hunk 2 above)
Comment (webknjaz) [reply pyproject.toml:301]: This is needed to measure coverage outside `src/`. One setting can only contain importable names while the other can have both importables and paths.

I had to do a lot of experiments to arrive to this config shape. Other variants may output incorrect paths into the XML report, confusing Codecov.

File: tests/cython/run_test_cython.py (hunk 3 above)
Comment (webknjaz) [reply tests/cython/run_test_cython.py:1]: Yeah, I thought so. But it was late and I was lazy :)

Will do!

File: .github/workflows/ci.yml (hunk 4 above)
Comment (webknjaz) [reply .github/workflows/ci.yml:260]: No, they just use different uploader and APIs. But newer versions are less prone to flaky upload crashes for some reason, as I noticed.

File: pyproject.toml (hunk 1 above)
Comment (webknjaz) [reply pyproject.toml:279]: @jakkdl this is not Codecov but coveragepy.

File: pyproject.toml (hunk 2 above)
Comment (webknjaz) [reply pyproject.toml:301]: (this also allows including that Cython smoke test)

**Comment by webknjaz** [comment]
pre-commit.ci run

File: pyproject.toml (hunk 2 above)
Comment (graingert) [reply pyproject.toml:301]: Probably that smoke test should be inside src/trio/_test

**Review by graingert — CHANGES_REQUESTED** [review]
Let's use source, or source_pkgs not both

File: pyproject.toml (hunk 2 above)
Comment (graingert) [reply pyproject.toml:301]: Alternatively we could use `source_pkgs = ["trio", "tests"]`

**Review by graingert — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 2 above)
Comment (webknjaz) [reply pyproject.toml:301]: @graingert this won't work well. The combination of the two is the only config that is known not to cause problems one way or the other. Plus, it represents the shape of the project exactly — the source is the entire Git repo; the importable is whatever is in `site-packages` but it's also mapped back to the source for proper representation within both coveragepy and codecov.

I've seen cases where it would display `test_*.py` files in the root of the repository in Codecov while showing the correct paths in the coveragepy's own HTML output. And that's just one example of how things tend to break.

Can we please rely on my experience here? I happened to maintain the use of coveragepy specifically a lot in the past few years, and would like to avoid weird behaviors just because this project wants to deviate from what's gone through countless hours of debugging rather surprising corner cases.

**Comment by webknjaz** [comment]
Merge queue didn't work for the first time because of the flakiness of the Codecov v3 action, FTR.

I've added it back. As for the test relocation and other suggestions, they are out of the scope because I don't want to fit in more requirements into this effort. Those changes can always be tried out separately and shouldn't be blocking — and it would be easier to see if they work in isolation, since this already combines a bit more changes than I'd like in one PR.

File: pyproject.toml (hunk 2 above)
Comment (graingert) [reply pyproject.toml:301]: I added the source_pkgs feature to coveragepy for this use case

Comment (webknjaz) [reply pyproject.toml:301]: @graingert really? I didn't know!

Still, there were cases where this specific configuration layout was the only thing that made it possible for Codecov not to break in some cases. This was in both src- and flat-layout projects (some in aio-libs, some elsewhere).

Though, I feel like it's still important to cover other things that might appear outside `src/`. Sometimes, these are helper scripts or something similar — by collecting such coverage, it is possible to configure Codecov-reported statuses better and be more conscious about what to take into account through exclusions.

As for adding `"tests"` into `source_pkgs`, that might work, but would not address the “covering everything” part. Plus, I'm worried about the side effects it may have on Codecov.
